pipeline_server/src/main.rs

```diff
@@ -0,0 +1,873 @@
+        .unwrap_or_else(|e| http_resp_from_error(&e))
+}
+
+/// `/cancel_project` request parameters.
+#[derive(Deserialize)]
+struct CancelProjectRequest {
+    /// Project id.
+    project_id: ProjectId,
+    /// Latest project version known to the client.
+    version: Version,
+}
+
+/// Cancel outstanding compilation request.
+///
+/// The client should poll the `/project_status` endpoint
+/// to determine when the cancelation request completes.
+///
+/// # HTTP errors
+///
+/// * `NOT_FOUND` - `project_id` does not exist in the database.
+///
+/// * `CONFLICT` - project version specified in the request doesn't match the
+///   latest project version in the database.
+#[post("/cancel_project")]
+async fn cancel_project(
+    state: WebData<ServerState>,
+    request: web::Json<CancelProjectRequest>,
+) -> impl Responder {
+    state
+        .db
+        .lock()
+        .await
+        .cancel_project(request.project_id, request.version)
+        .await
+        .map(|_| HttpResponse::Ok().finish())
+        .unwrap_or_else(|e| http_resp_from_error(&e))
+}
+
+/// `/delete_project` request parameters.
+#[derive(Deserialize)]
+struct DeleteProjectRequest {
+    project_id: ProjectId,
+}
+
+/// Delete a project.
+///
+/// # HTTP errors
+///
+/// * `BAD_REQUEST` - the project has one or more running pipelines.
+///
+/// * `NOT_FOUND` - `project_id` does not exist in the database.
+#[post("/delete_project")]
+async fn delete_project(
+    state: WebData<ServerState>,
+    request: web::Json<DeleteProjectRequest>,
+) -> impl Responder {
+    let db = state.db.lock().await;
+
+    match db.list_project_pipelines(request.project_id).await {
+        Ok(pipelines) => {
+            if pipelines.iter().any(|pipeline| !pipeline.killed) {
+                return HttpResponse::BadRequest()
+                    .body("Cannot delete a project while some of its pipelines are running");
+            }
+        }
+        Err(e) => return http_resp_from_error(&e),
+    }
+
+    db.delete_project(request.project_id)
+        .await
+        .map(|_| HttpResponse::Ok().finish())
+        .unwrap_or_else(|e| http_resp_from_error(&e))
+}
+
+/// `/new_config` request parameters.
+#[derive(Deserialize)]
+struct NewConfigRequest {
+    /// Project to create config for.
+    project_id: ProjectId,
+    /// Config name.
+    name: String,
+    /// YAML code for the config.
+    config: String,
+}
+
+/// Response to a `/new_config` request.
+#[derive(Serialize)]
+struct NewConfigResponse {
+    /// Id of the newly created config.
+    config_id: ConfigId,
+    /// Initial config version (this field is always set to 1).
+    version: Version,
+}
+
+/// Create a new project configuration.
+///
+/// # HTTP errors
+///
+/// * `NOT_FOUND` - `project_id` does not exist in the database.
+#[post("/new_config")]
+async fn new_config(
+    state: WebData<ServerState>,
+    request: web::Json<NewConfigRequest>,
+) -> impl Responder {
```

Comment:
I wonder whether this code could be simplified.
A controller can have a set of apis with input/output data structures.
the web server just sends the data to the controller and serializes back the data structure or the error.
same structure for all requests as long as the data structures implement some trait.

Reply:
And the fact that the state is in a DB should also be abstracted by the controller.